sync/class.jetpack-sync-module-import.php

```diff
@@ -1,87 +1,143 @@
 
-		// Movable type, RSS, Livejournal
-		add_action( 'import_done', array( $this, 'sync_import_done' ) );
+		// WordPress.
+		add_action( 'import_start', array( $this, 'sync_import_action' ) );
+
+		// Movable type, RSS, Livejournal.
+		add_action( 'import_done', array( $this, 'sync_import_action' ) );
 
-		// WordPress, Blogger, Livejournal, woo tax rate
-		add_action( 'import_end', array( $this, 'sync_import_end' ) );
+		// WordPress, Blogger, Livejournal, woo tax rate.
+		add_action( 'import_end', array( $this, 'sync_import_action' ) );
 	}
 
 	public function set_defaults() {
-		$this->import_end = false;
+		$this->synced_actions = array();
 	}
 
-	public function sync_import_done( $importer ) {
-		// We already ran an send the import
-		if ( $this->import_end ) {
-			return;
-		}
-
-		$importer_name = $this->get_importer_name( $importer );
-
-		/**
-		 * Sync Event that tells that the import is finished
-		 *
-		 * @since 5.0.0
-		 *
-		 * $param string $importer
-		 */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-		$this->import_end = true;
-	}
+	public function sync_import_action( $importer ) {
+		$import_action = current_filter();
+		// Map action to event name.
+		$sync_action = self::$import_sync_action_map[ $import_action ];
 
-	public function sync_import_end() {
-		// We already ran an send the import
-		if ( $this->import_end ) {
+		// Only sync each action once per import.
+		if ( array_key_exists( $sync_action, $this->synced_actions ) && $this->synced_actions[ $sync_action ] ) {
 			return;
 		}
 
-		$this->import_end = true;
-		$importer         = 'unknown';
-		$backtrace        = wp_debug_backtrace_summary( null, 0, false );
-		if ( $this->is_importer( $backtrace, 'Blogger_Importer' ) ) {
-			$importer = 'blogger';
-		}
+		// Mark this action as synced.
+		$this->synced_actions[ $sync_action ] = true;
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WC_Tax_Rate_Importer' ) ) {
-			$importer = 'woo-tax-rate';
+		// Prefer self-reported $importer value.
+		if ( ! $importer ) {
+			// Fall back to inferring by calling class name.
+			$importer = self::get_calling_importer_class();
 		}
 
-		if ( 'unknown' === $importer && $this->is_importer( $backtrace, 'WP_Import' ) ) {
-			// phpcs:ignore WordPress.WP.CapitalPDangit
-			$importer = 'wordpress';
+		// Get $importer from known_importers.
+		$known_importers = Jetpack_Sync_Settings::get_setting( 'known_importers' );
+		if ( isset( $known_importers[ $importer ] ) ) {
+			$importer = $known_importers[ $importer ];
 		}
 
 		$importer_name = $this->get_importer_name( $importer );
 
-		/** This filter is already documented in sync/class.jetpack-sync-module-posts.php */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
+		// $sync_action should be either 'jetpack_sync_import_start' or 'jetpack_sync_import_end' as mapped via $import_sync_action_map.
+		do_action( $sync_action, $importer, $importer_name );
```

Comment:
I think this is going to trip up our developer.jetpack.com parser, or at least not be able to parse it.

@jeherve Do you have ideas of how to document this? We could add the two `do_action` calls in another file that is never called for the sake of documentation, linking back to here as the actual place it fires? I'm okay adding that after merging to not block this so we can get to freeze.

Reply:
Since there are only 2 possible `$sync_action`s, we could also just use a conditional or a switch statement here...

```
switch( $sync_action ) {
	case 'jetpack_sync_import_start':
		do_action( 'jetpack_sync_import_start' ... );
		break;
	...
}
```

Reply:
Yeah, let's go with a conditional. It's not super pretty, but it should make the parser happy and may actually end up being easier to get back to later, when looking for specific actions for example.

Reply:
Thank you both. Let's do the switch and add inline documentation above each do_action. @jonathansadowski do you have bandwidth to knock that out now-ish? If not, I can get to it in a short bit.

Reply:
Yep, grabbing it right now. Thanks @jeherve & @kraftbj.